### PR TITLE
2173 let cloud engineers able to join any capability without approval

### PR DESCRIPTION
**Pull request by paulseghers**
closes dfds/cloudplatform/issues/2173

# Additional Review Notes
Proper reloading of components as a response to certain actions was also handled/explored in this task.
I will squash and merge I promise

### CODE REVIEW

File: src/src/SelfServiceApiClient.js [hunk 1]
```diff
@@ -283,7 +283,19 @@ export class SelfServiceApiClient {
     this.responseHandler(response);
 
     if (!response.ok) {
-      console.log("response was: ", response.status);
+      console.log(`for url ${url} response was: ${response.status}`);
```
Comment (andreasfrisch) [inline src/src/SelfServiceApiClient.js:286]: Do we want shotgun debug logs?
```suggestion

```

File: src/src/SelfServiceApiClient.js [hunk 2]
```diff
@@ -283,7 +283,19 @@ export class SelfServiceApiClient {
...
+    return response;
+  }
+
+  async requestWithToken(url, method = "GET", payload = null) {
+    const accessToken = await getSelfServiceAccessToken();
+    const response = await callApi(url, accessToken, method, payload);
+    this.responseHandler(response);
+
+    if (!response.ok) {
+      console.log(`for url ${url} response was: ${response.status}`);
```
Comment (andreasfrisch) [inline src/src/SelfServiceApiClient.js:298]: Do we want shotgun debug logs?
```suggestion

```

File: src/src/SelfServiceApiClient.js [hunk 3]
```diff
@@ -283,7 +283,19 @@ export class SelfServiceApiClient {
...
 
     if (!response.ok) {
-      console.log("response was: ", response.status);
+      console.log(`for url ${url} response was: ${response.status}`);
+      return undefined;
+    }
+    return response;
+  }
+
+  async requestWithToken(url, method = "GET", payload = null) {
```
Comment (andreasfrisch) [inline src/src/SelfServiceApiClient.js:292]: We have  a `fetchWithToken` function as well.

They seem identical except for the name and the default parameters.
I think we need only one.

File: src/src/SelfServiceApiClient.js [hunk 4]
```diff
@@ -587,6 +599,31 @@ export class SelfServiceApiClient {
       );
     }
   }
+
+  CheckCanBypassMembershipApproval(capabilityDefinition) {
```
Comment (kralle333) [inline src/src/SelfServiceApiClient.js:603]: lower case start

File: src/src/SelfServiceApiClient.js [hunk 5]
```diff
@@ -587,6 +599,31 @@ export class SelfServiceApiClient {
...
+      throw Error(
+        "Error! No join link found for capability " +
+          capabilityDefinition.capabilityId +
+          ", or user not allowed to join directly",
+      );
+    }
+    return link;
+  }
+
+  async ByPassMembershipApproval(capabilitydefinition) {
```
Comment (kralle333) [inline src/src/SelfServiceApiClient.js:615]: lower case start

File: src/src/SelfServiceApiClient.js [hunk 6]
```diff
@@ -587,6 +599,31 @@ export class SelfServiceApiClient {
...
+
+  CheckCanBypassMembershipApproval(capabilityDefinition) {
+    const link = capabilityDefinition?._links?.joinCapability;
+    if (!link || !link.allow.includes("POST")) {
+      throw Error(
+        "Error! No join link found for capability " +
+          capabilityDefinition.capabilityId +
+          ", or user not allowed to join directly",
+      );
+    }
```
Comment (kralle333) [inline src/src/SelfServiceApiClient.js:611]: you talked about better error handling here

File: src/src/SelfServiceApiClient.js (hunk 4 above)
Comment (kralle333) [reply src/src/SelfServiceApiClient.js:603]: bypass one word

File: src/src/SelfServiceApiClient.js (hunk 5 above)
Comment (kralle333) [reply src/src/SelfServiceApiClient.js:615]: bypass one word

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 7]
```diff
@@ -48,10 +50,11 @@ function SelectedCapabilityProvider({ children }) {
   const [kafkaClusters, setKafkaClusters] = useState([]);
   const [selectedKafkaTopic, setSelectedKafkaTopic] = useState(null);
   const [membershipApplications, setMembershipApplications] = useState([]);
-  const [leaveCapability, setLeaveCapability] = useState([]);
+  //const [leaveCapability, setLeaveCapability] = useState([]);
```
Comment (kralle333) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:53]: why?

File: src/src/SelfServiceApiClient.js [hunk 8]
```diff
@@ -587,6 +599,31 @@ export class SelfServiceApiClient {
       );
     }
   }
+
+  CheckCanBypassMembershipApproval(capabilityDefinition) {
+    const link = capabilityDefinition?._links?.joinCapability;
```
Comment (andreasfrisch) [inline src/src/SelfServiceApiClient.js:604]: 'join' and 'bypassMembershipApplication' does not feel like they relate.
I know they do, but it feels a bit strange

File: src/src/SelfServiceApiClient.js (hunk 5 above)
Comment (andreasfrisch) [inline src/src/SelfServiceApiClient.js:615]: Do we really want multiple functions for this?

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 9]
```diff
@@ -330,6 +342,11 @@ function SelectedCapabilityProvider({ children }) {
     await selfServiceApiClient.submitCancelDeleteCapability(details);
   }, [details]);
 
+  const ByPassMembershipApproval = async () => {
+    await selfServiceApiClient.ByPassMembershipApproval(details);
```
Comment (andreasfrisch) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:346]: This one can throw a bunch of errors; How are they handled?

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 10]
```diff
@@ -358,6 +375,17 @@ function SelectedCapabilityProvider({ children }) {
...
   }, [isLoadedMembers, membersList]);
 
+  useEffect(() => {
+    if (details) {
+      loadMembershipApplications();
+      getBypassMembershipApplication();
+    } else {
+      setMembers([]);
+      setMembershipApplications([]);
+      setKafkaClusters([]);
```
Comment (kralle333) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:385]: do you need to hide other things besides applications?

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 11]
```diff
@@ -384,6 +412,8 @@ function SelectedCapabilityProvider({ children }) {
     }
   }, [awsAccountRequested]);
 
+  useEffect(() => {}, []);
```
Comment (kralle333) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:415]: ?

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 12]
```diff
@@ -90,6 +94,13 @@ function SelectedCapabilityProvider({ children }) {
     kafkaClusterTopicList();
   }, [clustersList]);
 
+  // load bypass rights:
+  const getBypassMembershipApplication = useCallback(async () => {
+    const result =
+      await selfServiceApiClient.CheckCanBypassMembershipApproval(details);
```
Comment (andreasfrisch) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:100]: This is not asynchronous, is it?
Did you want to call `BypassMembershipApproval`

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 13]
```diff
@@ -358,6 +375,17 @@ function SelectedCapabilityProvider({ children }) {
     }
   }, [isLoadedMembers, membersList]);
 
+  useEffect(() => {
+    if (details) {
+      loadMembershipApplications();
+      getBypassMembershipApplication();
```
Comment (kralle333) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:381]: bad naming: you are not only getting, you are first fetching and then setting a variable

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 14]
```diff
@@ -419,6 +448,8 @@ function SelectedCapabilityProvider({ children }) {
     isDeleted,
     updateDeletionStatus,
     showCosts,
+    canBypassMembershipApplication,
+    ByPassMembershipApproval,
```
Comment (kralle333) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:452]: bypass one word, lowercase b

File: src/src/pages/capabilities/summary/index.js [hunk 15]
```diff
@@ -34,6 +41,7 @@ function JoinDialog({ name, isSubmitting, onCloseRequested, onSubmitClicked }) {
   return (
     <>
       <Modal
+        style={{ position: "absolute" }}
```
Comment (kralle333) [inline src/src/pages/capabilities/summary/index.js:44]: you need this?

File: src/src/pages/capabilities/SelectedCapabilityContext.js (hunk 11 above)
Comment (andreasfrisch) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:415]: What is the purpose of this?
```suggestion

```

File: src/src/pages/capabilities/summary/index.js [hunk 16]
```diff
@@ -54,6 +62,16 @@ function JoinDialog({ name, isSubmitting, onCloseRequested, onSubmitClicked }) {
...
           </i>
         </Text>
+        {canBypassMembershipApplications && (
+          <Button
+            variation="danger"
+            style={{ position: "absolute", bottom: "1rem" }}
+            disabled={isSubmitting}
+            onClick={onBypassClicked}
+          >
+            FORCE JOIN (CE)
```
Comment (kralle333) [inline src/src/pages/capabilities/summary/index.js:72]: think this is a bit overkill to call it FORCE JOIN: endpoint is called /join and is valid for you

**Review by andreasfrisch — CHANGES_REQUESTED** [review]
There is some cleaning left to be done, I believe.
Otherwise it looks good :)

File: src/src/hooks/Capabilities.js [hunk 17]
```diff
@@ -1,6 +1,7 @@
 import { useEffect, useState } from "react";
 import { useSelfServiceRequest } from "./SelfServiceApi";
 import { getAnotherUserProfilePictureUrl } from "../GraphApiClient";
+import { set } from "date-fns";
```
Comment (kralle333) [inline src/src/hooks/Capabilities.js:4]: do you need this?

File: src/src/hooks/Capabilities.js [hunk 18]
```diff
@@ -1,6 +1,7 @@
...
 
 export function useCapabilities() {
   // const { errorMessage } = useSelfServiceRequest();
```
Comment (kralle333) [inline src/src/hooks/Capabilities.js:7]: remove

File: src/src/SelfServiceApiClient.js (hunk 3 above)
Comment (kralle333) [inline src/src/SelfServiceApiClient.js:292]: remove the other similar function `fetchWithToken`

**Review by kralle333 — CHANGES_REQUESTED** [review]
check da comments.

Would be nice if you did a review yourself before assigning

File: src/src/pages/capabilities/summary/index.js (hunk 16 above)
Comment (andreasfrisch) [reply src/src/pages/capabilities/summary/index.js:72]: This was my original suggestion:
In most command line interfaces the '--force' flag means 'ignore usual defenses and do it anyway'.
That is exactly what we are doing here.

The word force also emphasizes that it is not the normal approach and sets it apart from the normal join button that we _just_ pressed.

On the other hand, I also argued for the join button automatically approving memberships for CEs, so I might be completely out of my mind 🤷‍♂️

File: src/src/pages/capabilities/SelectedCapabilityContext.js [hunk 19]
```diff
@@ -90,6 +94,13 @@ function SelectedCapabilityProvider({ children }) {
     kafkaClusterTopicList();
   }, [clustersList]);
 
+  // load bypass rights:
+  const getBypassMembershipApplication = useCallback(async () => {
```
Comment (andreasfrisch) [inline src/src/pages/capabilities/SelectedCapabilityContext.js:98]: On second read-through; Why don't we utilize our generic approach to sending requests to the server and updating a variable upon completion. That is exactly what you are doing here, is it not?

The `useSelfServiceRequest` should suffice?

File: src/src/pages/capabilities/summary/index.js (hunk 16 above)
Comment (paulseghers) [reply src/src/pages/capabilities/summary/index.js:72]: If I recall well, Mikkel and Audun asked for a separate join button for CEs
I have no opinion on this.

**Review by andreasfrisch — APPROVED** [review]

**Pull request merged**